python/tutorials/10-block-scaled-matmul.py

```diff
@@ -0,0 +1,335 @@
+and iteration steps.
+
+.. code-block:: bash
+
+    # FP4
+    python 10-block-scaled-matmul.py --format nvfp4
+    python 10-block-scaled-matmul.py --format mxfp4 --K_range 512 8192 --bench
+
+    # FP8
+    python 10-block-scaled-matmul.py --format mxfp8 --K_range 8192 16384 --K_step 2048 --bench
+
+Future updates to this tutorial which support mixed precision block scaled matmul are planned.
+"""
+
+# %%
+# Background
+# ----------
+#
+# CUDA devices that support PTX 8.7 and later can utlize block scaled matrix multiply
+# instructions. In order for low latency access to these scale factors in the fast
+# inner loop over tensor core MMAs, it is important to ensure that the blocked
+# scale factors are stored in a contiguous memory layout according to their access
+# pattern.
+#
+# The block scaled matmul tensor core instructions compute the following product:
+#
+#     C = (A * scale_a) @ (B * scale_b)
+#
+# where scale_a and scale_b are the blocked scale factors for the A and B matrices.
+# Under block scaled matmul, each scale factor is broadcast and multiplied across a
+# vector of elements from the A and B matrices, usually along their respective K axes.
+# The number of elements of A and B over which each scale factor is broadcast is herein
+# refered to as the vector size (VEC_SIZE).
+#
+# In a linear row-major layout, the scale factors would take the shape
+#
+#     (M, K // VEC_SIZE) and (N, K // VEC_SIZE)   [1]
+#
+# in global memory. However, to avoid non-contiguous memory access, it is beneficial to
+# instead store the scale factors in a packed block layout. For the LHS matrix this layout
+# is given by
+#
+#     (M // 32 // 4, K // VEC_SIZE // 4, 32, 4, 4)   [2].
+#
+# In this way, each tensor core MMA in the fast inner loop over K blocks can achieve contiguous
+# access of a block of 128 rows of scale factors along the M axis, for each BLOCK_M x BLOCK_K
+# subtile of the matrix A.
+#
+# In order to conform with Triton's language semantics for dot_scaled, the scale factors
+# are prepared in the above 5D layout [2], but are then logically transposed and reshaped into
+# the 2D layout [1]. These ops are lowered into a Tensor Memory (TMEM) copy operation
+# which moves the scale factors directly from SMEM to TMEM prior to consumption by the tcgen05.mma operation.
+#
+# For more detailed information, see
+# 1. The scale factor layout discussion at https://docs.nvidia.com/cuda/parallel-thread-execution/#tcgen05-mma-scale-factor-a-layout-1x.
+# 2. The TMEM copy operation semantics at https://docs.nvidia.com/cuda/parallel-thread-execution/#tensorcore-5th-generation-instructions-tcgen05-cp
+#
+
+import argparse
+
+import torch
+import triton
+import triton.language as tl
+import triton.tools.experimental_descriptor
+import triton.profiler as proton
+from triton.tools.experimental_descriptor import TmaDescKernelParam
+from triton._internal_testing import MXFP4Tensor, MXScaleTensor
```

Comment:
imo they are useful enough to be put into `triton.tools.mxfp`

Reply:
depending on testing in the tutorial is a bit odd indeed. I agree isolating useful helpers to put into `tools` would be good

Reply:
Thanks, I didn't originally want to presume they are sufficiently useful to warrant use as a tool, but happy to make the change; updated.